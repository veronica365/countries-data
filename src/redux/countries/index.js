import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';
import axios from 'axios';
import { API } from '../../constants';

const findItem = (id, countries) => countries.filter((country) => String(country.id) === id);
const searchItem = (param, countries) => countries.filter(
  (country) => String(country.id) === param
      || String(country.name).toLowerCase().includes(String(param).toLowerCase())
      || String(country.currency).toLowerCase().includes(String(param).toLowerCase())
      || String(country.capital).toLowerCase().includes(String(param).toLowerCase())
      || String(country.region).toLowerCase().includes(String(param).toLowerCase())
      || String(country.currency_name).toLowerCase().includes(String(param).toLowerCase())
      || String(country.subregion).toLowerCase().includes(String(param).toLowerCase()),
);

export const getData = createAsyncThunk('get', async (_, thunkAPI) => {
  try {
    const { countries } = thunkAPI.getState().countries;
    if (countries.length) return countries;
    const { data } = await axios.get(`${API}`);
    return data;
  } catch (error) {
    return thunkAPI.rejectWithValue('Something went wrong');
  }
});
export const findCountry = createAsyncThunk('find', async (id, thunkAPI) => {
  try {
    const { countries } = thunkAPI.getState().countries;
    if (countries.length) return findItem(id, countries);
    const { data } = await axios.get(`${API}`);
    return findItem(id, data);
  } catch (error) {
    return thunkAPI.rejectWithValue('Something went wrong');
  }
});
export const searchCountry = createAsyncThunk(
  'search',
  async (param, thunkAPI) => {
    try {
      const { countries } = thunkAPI.getState().countries;
      if (countries.length) return searchItem(param, countries);
      const { data } = await axios.get(`${API}`);
      return searchItem(param, data);
    } catch (error) {
      return thunkAPI.rejectWithValue('Something went wrong');
    }
  },
);

const countrieslice = createSlice({
  name: 'countries',
  initialState: {
    countries: [],
    countriesBySearch: [],
    countryById: {},
    isLoading: false,
    error: null,
  },
  reducers: {
    setCountryById: (state, action) => ({
      ...state,
      countryById: action.payload,
      countriesBySearch: [],
      isLoading: false,
    }),
  },
  extraReducers: (builder) => {
    builder.addCase(getData.pending, (state) => ({
      ...state,
      isLoading: true,
    }));
    builder.addCase(getData.fulfilled, (state, action) => ({
      ...state,
      countries: action.payload,
      isLoading: false,
    }));
    builder.addCase(getData.rejected, (state, action) => ({
      ...state,
      error: action.payload,
      isLoading: false,
    }));
    builder.addCase(findCountry.pending, (state) => ({
      ...state,
      isLoading: true,
    }));
    builder.addCase(findCountry.fulfilled, (state, action) => ({
      ...state,
      countryById: action.payload[0] || {},
      isLoading: false,
    }));
    builder.addCase(findCountry.rejected, (state, action) => ({
      ...state,
      error: action.payload,
      isLoading: false,
    }));
    builder.addCase(searchCountry.pending, (state) => ({
      ...state,
      isLoading: true,
    }));
    builder.addCase(searchCountry.fulfilled, (state, action) => ({
      ...state,
      countriesBySearch: action.payload,
      isLoading: false,
    }));
    builder.addCase(searchCountry.rejected, (state, action) => ({
      ...state,
      error: action.payload,
      isLoading: false,
    }));
  },
});

export const { setCountryById } = countrieslice.actions;

export default countrieslice.reducer;
