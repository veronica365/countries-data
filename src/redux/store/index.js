import { configureStore } from '@reduxjs/toolkit';
import logger from 'redux-logger';
import countryReducer from '../countries';

const store = configureStore({
  reducer: {
    countries: countryReducer,
  },
  middleware: (getDefaultMiddleware) => getDefaultMiddleware().concat(logger),
  devTools: process.env.NODE_ENV !== 'production',
});

export default store;
