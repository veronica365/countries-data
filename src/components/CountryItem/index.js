import PropTypes from 'prop-types';
import { useDispatch } from 'react-redux';
import { Link, useNavigate } from 'react-router-dom';
import { setCountryById } from '../../redux/countries';

export default function CountryItem({ country }) {
  const navigate = useNavigate();
  const dispatch = useDispatch();
  const clickHandler = (event) => {
    event.preventDefault();
    dispatch(setCountryById({}));
    navigate(`/country/${country.id}`);
  };
  return (
    <article className="country">
      <Link to={`/country/${country.id}`} onClick={clickHandler}>
        <div className="country-info">
          <h2 className="country-name">{country.name}</h2>
          <div className="country-data">
            <span className="country-data-value">
              <i className="fa fa-mobile" />
              {country.phone_code}
            </span>
            <span className="country-data-value">
              <i className="fa fa-money" />
              {country.currency}
            </span>
            <span className="country-data-value">
              <i className="fa fa-map-marker" />
              {parseFloat(country.latitude).toFixed(2)}
              {' '}
&nbsp;&nbsp;
              {parseFloat(country.longitude).toFixed(2)}
            </span>
          </div>
        </div>
      </Link>
    </article>
  );
}

CountryItem.propTypes = {
  country: PropTypes.shape({
    id: PropTypes.number.isRequired,
    name: PropTypes.string.isRequired,
    phone_code: PropTypes.string.isRequired,
    currency: PropTypes.string.isRequired,
    latitude: PropTypes.string.isRequired,
    longitude: PropTypes.string.isRequired,
  }).isRequired,
};
