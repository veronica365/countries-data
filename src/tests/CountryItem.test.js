import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';
import { render } from '@testing-library/react';
import CountryItem from '../components/CountryItem';
import { lessCountryData } from './__Mocks__/countries';
import store from '../redux/store';

describe('CountryItem', () => {
  it('should render with fewer country details as props', () => {
    const { container } = render(
      <Provider store={store}>
        <BrowserRouter>
          <CountryItem country={lessCountryData} />
        </BrowserRouter>
      </Provider>,
    );
    expect(container).toMatchSnapshot();
  });
});
