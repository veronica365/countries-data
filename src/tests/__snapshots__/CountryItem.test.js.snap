// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CountryItem should render with fewer country details as props 1`] = `
<div>
  <article
    class="country"
  >
    <a
      href="/country/1"
    >
      <div
        class="country-info"
      >
        <h2
          class="country-name"
        >
          Afghanistan
        </h2>
        <span
          class="country-flag"
        >
          ðŸ‡¦ðŸ‡«
        </span>
        <div
          class="country-data"
        >
          <span
            class="country-data-value"
          >
            <i
              class="fa fa-mobile"
            />
            93
          </span>
          <span
            class="country-data-value"
          >
            <i
              class="fa fa-money"
            />
            AFN
          </span>
          <span
            class="country-data-value"
          >
            <i
              class="fa fa-map-marker"
            />
            33.00
             
            Â Â 
            65.00
          </span>
        </div>
      </div>
    </a>
  </article>
</div>
`;
